### PR TITLE
Check AGPL code doesn't import enterprise

### PR DESCRIPTION
**Pull request by spikecurtis**
This adds a check to ensure we don't accidentally import enterprise code from AGPL code, which would cause "AGPL" builds to not really be AGPL.

Oh, and if you are wondering, "can ruleguard do this?," the answer is [no](https://github.com/quasilyte/go-ruleguard/issues/78)

### CODE REVIEW

File: scripts/check_enterprise_imports.sh [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+cdroot
+
+set +e
+find . -regex ".*\.go" | grep -v "./enterprise" | xargs grep -n "github.com/coder/coder/enterprise"
+# reverse the exit code because we want this script to fail if grep finds anything.
+status=$?
+set -e
+if [ $status -eq 0 ]; then
+	echo "AGPL code cannot import enterprise!"
+	exit 1
```
Comment (deansheather) [inline scripts/check_enterprise_imports.sh:18]:
```suggestion
	error "AGPL code cannot import enterprise!"
```

File: scripts/check_enterprise_imports.sh [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+fi
+echo "AGPL imports OK"
+exit 0
```
Comment (deansheather) [inline scripts/check_enterprise_imports.sh:21]:
```suggestion
log "AGPL imports OK"
```

**Review by deansheather — APPROVED** [review]

File: scripts/check_enterprise_imports.sh (hunk 1 above)
Comment (spikecurtis) [reply scripts/check_enterprise_imports.sh]: clearly my bash is pretty weak.  Thanks.

File: .github/workflows/coder.yaml [hunk 3]
```diff
@@ -102,6 +102,15 @@ jobs:
...
 
+  check-enterprise-imports:
+    name: check/enterprise-imports
+    timeout-minutes: 5
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Check imports of enterprise code
+        run: ./scripts/check_enterprise_imports.sh
+
```
Comment (f0ssel) [inline .github/workflows/coder.yaml:113]: Would it be OK to just let the lint step handle this?

**Review by f0ssel — APPROVED** [review]

File: .github/workflows/coder.yaml (hunk 3 above)
Comment (spikecurtis) [reply .github/workflows/coder.yaml:113]: I don't think the lint Action calls `make lint/go`, it directly calls golangci-lint

File: scripts/check_enterprise_imports.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+# This file checks all our AGPL licensed source files to be sure they don't
+# import any enterprise licensed packages (the inverse is fine).
+
+set -euo pipefail
+# shellcheck source=scripts/lib.sh
+source "$(dirname "${BASH_SOURCE[0]}")/lib.sh"
+cdroot
+
+set +e
+find . -regex ".*\.go" | grep -v "./enterprise" | xargs grep -n "github.com/coder/coder/enterprise"
```
Comment (mafredri) [inline scripts/check_enterprise_imports.sh:12]:
```suggestion
find . -name ./enterprise -prune -o -name '*.go' -print0 | xargs -0 grep -n "github.com/coder/coder/enterprise"
```

Not at a computer; so can’t test. But use if you want. Silly tool optimization. This even handles spaces in names which isn’t ever going to be a problem with Go source files 🤪

**Review by mafredri — COMMENTED** [review]
Looks like I was too slow with my review 😄.

Btw, I looked at the linked issue; this last comment makes it sound like it would be possible? https://github.com/quasilyte/go-ruleguard/issues/78#issuecomment-1103194156

**Comment by spikecurtis** [comment]
@mafredri it is possible to match on file names, which is one requirement, but it is not possible to match the `import` statements, which is the other requirement.

**Pull request merged**